src/NuGet.Clients/NuGet.CommandLine/MsBuildToolsetEx.cs

```diff
@@ -0,0 +1,128 @@
+        {
+            _toolset = toolset;
+            _installDateTime = installDateTime;
+        }
+
+        internal MsBuildToolsetEx(ISetupInstance sxsToolset)
+        {
+            _toolset = new Toolset(toolsVersion: sxsToolset.GetInstallationVersion(),
+                        toolsPath: sxsToolset.GetInstallationPath(),
+                        projectCollection: null, 
+                        msbuildOverrideTasksPath: string.Empty);
+            _installDateTime = ConvertFILETIMEToDateTime(sxsToolset.GetInstallDate());
+        }
+
+        public string ToolsVersion
+        {
+            get
+            {
+                return _toolset?.ToolsVersion;
+            }
+        }
+
+        public Version ParsedToolsVersion
+        {
+            get
+            {
+                if (_version == null)
+                {
+                    if (!Version.TryParse(_toolset?.ToolsVersion, out _version))
+                    {
+                        _version = new Version(0, 0);
+                    }
+                }
+
+                return _version;
+            }
+        }
+
+        private static Version SafeParseVersion(string version)
+        {
+            Version result;
+
+            if (Version.TryParse(version, out result))
+            {
+                return result;
+            }
+            else
+            {
+                return new Version(0, 0);
+            }
+        }
+
+        public string ToolsPath
+        {
+            get
+            {
+                return _toolset?.ToolsPath;
+            }
+        }
+
+        public static IEnumerable<MsBuildToolsetEx> AsMsToolsetExCollection(IEnumerable<Toolset> toolsets)
+        {
+            if (toolsets == null)
+            {
+                yield break;
+            }
+
+            foreach(var toolset in toolsets)
+            {
+                yield return new MsBuildToolsetEx(toolset);
+            }
+        }
+
+        private static DateTime ConvertFILETIMEToDateTime(FILETIME time)
+        {
+            long highBits = time.dwHighDateTime;
+            highBits = highBits << 32;
+            return DateTime.FromFileTimeUtc(highBits | (long)(uint)time.dwLowDateTime);
+        }
+
+        public int CompareTo(object obj)
+        {
+            var rhs = obj as MsBuildToolsetEx;
+            Debug.Assert(obj != null, "Comparison object not of correct type");
+            if (obj == null)
+            {
+                return 1;
+            }
+
+            // Compare versions
+            if (this.ParsedToolsVersion > rhs.ParsedToolsVersion)
+            {
+                return 1;
+            }
+            if (this.ParsedToolsVersion < rhs.ParsedToolsVersion)
+            {
+                return -1;
+            }
+
+            // Versions equal; compare by install date/time
+            return this._installDateTime > rhs._installDateTime ? 1 : -1;
```

Comment:
`MsBuildToolsetEx(Toolset toolset)` doesn't set the date. Consider initializing it with default `DateTime.MinValue`